src/mlib/path.h

```diff
@@ -76,31 +161,219 @@ mpath_parent (mstr_view path)
+
+/**
+ * @brief Obtain a relative path from the given filepath
+ *
+ * If the path has a root path, returns the content of the path following that
+ * root path, otherwise returns the same path itself.
+ */
+static inline mstr_view
+mpath_relative_path (mstr_view path, mpath_format f)
+{
+   mstr_view root = mpath_root_path (path, f);
+   return mstrv_subview (path, root.len, path.len);
+}
+
+/**
+ * @brief Convert the filepath from one format to a preferred form in another
+ * format.
+ *
+ * @note The return value must be freed with mstr_free()
+ */
+static inline mstr
+mpath_to_format (mpath_format from, mstr_view path, mpath_format to)
+{
+   mstr_mut ret = mstr_new (path.len);
+   const char *p = path.data;
+   char *out = ret.data;
+   const char *stop = path.data + path.len;
+   for (; p != stop; ++p, ++out) {
+      if (mpath_is_sep (*p, from)) {
+         *out = mpath_preferred_sep (to);
+      } else {
+         *out = *p;
+      }
+   }
+   return ret.mstr;
+}
+
+/**
+ * @brief Determine whether the given path is relative (not absolute)
+ */
+static inline bool
+mpath_is_relative (mstr_view path, mpath_format f)
+{
+   return !mpath_is_absolute (path, f);
+}
+
+/**
+ * @brief Convert the given path to an absolute path, if it is not already.
+ *
+ * @note The return value must be freed with mstr_free()
+ */
+static inline mstr
+mpath_absolute (mstr_view path, mpath_format f);
+
+/**
+ * @brief Resolve a path to an absolute path from the given base path.
+ *
+ * @note This is not the same as mpath_join(): If the given path is already
+ * absolute, returns that path unchanged. Otherwise, resolves that path as being
+ * relative to `base`.
+ *
+ * @note If `base` is also a relative path, it will also be given to
+ * mpath_absolute() to resolve it.
+ */
+static inline mstr
+mpath_absolute_from (mstr_view path, mstr_view base, mpath_format f)
+{
+   mstr_view rname = mpath_root_name (path, f);
+   mstr_view rdir = mpath_root_directory (path, f);
+   if (rname.len) {
+      if (rdir.len) {
+         // The path is already fully absolute
+         return mstr_copy (path);
+      } else {
+         mstr abs_base = mpath_absolute (base, f);
+         mstr_view base_rdir = mpath_root_directory (abs_base.view, f);
+         mstr_view base_relpath = mpath_relative_path (abs_base.view, f);
+         mstr_view relpath = mpath_relative_path (path, f);
+         mstr ret = mstr_copy (rname);
+         mstr_assign (&ret, mpath_join (ret.view, base_rdir, f));
+         mstr_assign (&ret, mpath_join (ret.view, base_relpath, f));
+         mstr_assign (&ret, mpath_join (ret.view, relpath, f));
+         mstr_free (abs_base);
+         return ret;
+      }
+   } else {
+      // No root name
+      if (rdir.len) {
+         if (f == MPATH_POSIX) {
+            // No root name, but a root directory on a POSIX path indicates an
+            // absolute path
+            return mstr_copy (path);
+         }
+         mstr abs_base = mpath_absolute (base, f);
```

Comment:
Note, this can result in a runtime recursive loop if `f == MPATH_WIN32` and is run on a POSIX machine.

I am not sure if that is a realistic concern. My understanding is that MPATH_NATIVE will only be used, but that selecting at runtime eases development.

```c
mpath_absolute_from (mstrv_lit("\\foo\\bar"), mstrv_lit ("blah"), MPATH_WIN32));
```

Reply:
I added a special case to work around this. Rather than crashing it will fall-back to the native format if `base` is non-absolute.